scripts/generate_mock_resolution_universe/main.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	f, err := lf.OpenLocalDepFile(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	g, err := io.Read(f)
+	if err != nil {
+		return err
+	}
+
+	// In-place updating doesn't actually check the client for the packages in the lockfile.
+	// For good measure, we do it here to load them into the cache.
+	group := &errgroup.Group{}
+	for _, n := range g.Nodes {
+		vk := n.Version
+		group.Go(func() error {
+			_, err := ddCl.Requirements(context.Background(), vk)
+			return err
+		})
+	}
+	_ = group.Wait()
+
+	_, err = remediation.ComputeInPlacePatches(context.Background(), cl, g, remediation.RemediationOptions{
+		DevDeps:    true,
+		MaxDepth:   -1,
+		AllowMajor: true,
+	})
+
+	return err
+}
+
+func getCachedVersions(cl *client.DepsDevClient) (map[resolve.PackageKey][]string, error) {
+	// Abuse the cache writing to get the list of encountered package versions.
+	cachePath := filepath.Join(os.TempDir(), "gr-cache")
+	if err := cl.WriteCache(cachePath); err != nil {
+		return nil, err
+	}
+	cacheFile := cachePath + ".resolve.deps"
+	defer os.Remove(cacheFile)
+	b, err := os.ReadFile(cacheFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var cache depsdevAPICache
+	dec := gob.NewDecoder(bytes.NewReader(b))
+	if err := dec.Decode(&cache); err != nil {
+		return nil, err
+	}
+
+	pkgVers := make(map[resolve.PackageKey][]string)
+	for vk := range cache.RequirementsCache {
+		pk := resolve.PackageKey{
+			Name:   vk.Name,
+			System: resolve.System(vk.System),
+		}
+		pkgVers[pk] = append(pkgVers[pk], vk.Version)
+	}
+
+	return pkgVers, nil
+}
+
+// Copy the relevant cache format from the depsdev_api_cache
+type depsdevAPICache struct {
+	RequirementsCache map[struct {
+		System  pb.System
+		Name    string
+		Version string
+	}][]byte
+}
+
+func (t *depsdevAPICache) GobDecode(b []byte) error {
+	type c depsdevAPICache
+	dec := gob.NewDecoder(bytes.NewReader(b))
+
+	return dec.Decode((*c)(t))
+}
+
+func makeUniverse(cl *client.DepsDevClient) (clienttest.ResolutionUniverse, error) {
+	pkgs, err := getCachedVersions(cl)
+	if err != nil {
+		return clienttest.ResolutionUniverse{}, err
+	}
+
+	pks := maps.Keys(pkgs)
+	slices.SortFunc(pks, func(a, b resolve.PackageKey) int { return a.Compare(b) })
+
+	if len(pks) == 0 {
+		return clienttest.ResolutionUniverse{}, errors.New("no packages found in cache")
+	}
+	// assume every package is the same system
+	system := pks[0].System
+
+	// Build the schema string.
+	schema := &strings.Builder{}
+	for _, pk := range pks {
+		vers := pkgs[pk]
+		slices.SortFunc(vers, system.Semver().Compare)
+		fmt.Fprintln(schema, pk.Name)
+		for _, v := range vers {
+			fmt.Fprintln(schema, "\t"+v)
+			reqs, err := cl.Requirements(context.Background(), resolve.VersionKey{
+				PackageKey:  pk,
+				Version:     v,
+				VersionType: resolve.Concrete,
+			})
+			if err != nil {
+				return clienttest.ResolutionUniverse{}, err
+			}
+			for _, r := range reqs {
+				// don't bother writing dev dependencies
```

Comment:
How about optional dependencies?

Reply:
npm tries to install optional dependencies if possible, so I'd like to keep them in the universe